[CSRNG] First Section of CSRNG Documentation (Overview Material) (#2463)
lowRISC/opentitan: hw/ip/csrng/doc/_index.md

---
title: "CSRNG HWIP Technical Specification"
---

# Overview

This document specifies the Cryptographically Secure Random Number Generator (CSRNG) hardware IP functionality.
Due to the importance of secure random number generation (RNG), it is a topic which is extensively covered in security standards.

> This IP targets compliance with both [BSI’s AIS31 recommendations for Common Criteria](https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Zertifizierung/Interpretationen/AIS_31_Functionality_classes_for_random_number_generators_e.pdf), as well as [NIST's SP 800-90](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf) (as referenced in [FIPS 140-3](https://csrc.nist.gov/publications/detail/fips/140/3/final)).

Author: Hi @imphil.  I'll always really impressed at how quickly you spot these, I really have to squint to note the difference. Do you always find these by eye?  (I'm also miffed that google Docs adds them).  Thanks.  I'll grep and remove them.

> This IP targets compliance with both [BSI’s AIS31 recommendations for Common Criteria](https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Zertifizierung/Interpretationen/AIS_31_Functionality_classes_for_random_number_generators_e.pdf), as well as [NIST's SP 800-90](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf) (as referenced in [FIPS 140-3](https://csrc.nist.gov/publications/detail/fips/140/3/final)).

Reviewer: Just years of pedantic proof reading :smile:

> This IP targets compliance with both [BSI's AIS31 recommendations for Common Criteria](https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Zertifizierung/Interpretationen/AIS_31_Functionality_classes_for_random_number_generators_e.pdf), as well as [NIST's SP 800-90](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf) (as referenced in [FIPS 140-3](https://csrc.nist.gov/publications/detail/fips/140/3/final)).

Reviewer 2: I think this ref should be to "NIST's SP 800-90A" rather than just 800-90

Author: Fixed. Thanks.

> Since these two standards use significantly different terminology, it is recommended that the reader refer to our RNG compliance strategy document for an overview of the various RNG classes and equivalencies between the two standards.

Reviewer: Do we have a link for the "compliance strategy document"?

Author: It's something that is linked currently in the gdoc version, however I have not yet migrated it to Markdown, though I intend to do so.   The simplest termporary path forward would be to include a URL to the gdoc for now.  Would that be okay you think?

Reviewer: Don't worry then, we can always add the link later.

Author: Ok.  Thanks.

The CSRNG IP supports both of these standards for both deterministic (DRNG) and true random number generation (TRNG).

> In NIST terms, it works with the entropy_source IP to satisfy the requirements as a DRBG or NRBG (non-deterministic random bit generator).

Reviewer: ```suggestion
In NIST terms, it works with the [Entropy Source IP]({{< relref "hw/ip/entropy_src/doc" >}}) to satisfy the requirements as a DRBG or NRBG (non-deterministic random bit generator).
```

(Please double-check if the link renders correctly.)

Author: Yes will do, thanks for the syntax help!

Author: Worked right the first time.  Very nice!

> In AIS31 language, this same implementation satisfies both the DRG.3 requirements for deterministic generation, or the PTG.3 requirements for cryptographically processes physical generation.

Author: It does satisfy both, but not simultaneously.  I think if I change the "both" to an "either" this will make it more clear.  Good catch though.

> This IP block is attached to the chip interconnect bus as a peripheral module conforming to the [Comportable guideline for peripheral functionality]({{< relref "doc/rm/comportability_specification" >}}), but also has direct hardware links to other IPs for secure internal transmission of random numbers.

Reviewer 2: maybe "secure internal transmission" -> "secure (non-software-observable) transmission" ?  or something like that.

Author: How about "secure and software-inaccessable transmission"?

Reviewer 2: SGTM.  I was mostly trying to get rid of "internal" since it's semi-ambiguous.

## Features

> - Together with a secure entropy source, provides support for both deterministic (DRNG) and true random number generation (TRNG).

Reviewer 3: what does secure mean here...? Do we mean it has a certain minimal bar..?

Author: There are a couple different meanings:
1. The seeds as they flow from the entropy source to the CSRNG are kept secret.  (seeds are transferred over a secure channel.)
2. There is a guarantee that the entropy level of the seeds meets some minimum entropy level (as defined by both the health checks and whitening implementation)  
Both of these points are discussed in SP 800-90B and AIS31.

Author: Hi @tjaychen, I've clarified this in the text. Please see the change in  https://github.com/lowRISC/opentitan/pull/2463/commits/d5f9ec4fc667f9d380a623f4173425a4c509e805

> - Compliant with NIST and BSI recommendations for random number generation

Reviewer: ```suggestion
- Compliant with NIST and BSI recommendations for random number generation.
```

- In deterministic mode, meets the requirements given in AIS31 for a DRG.3 class deterministic random number generator (DRNG) meaning it provides Forward Secrecy and Enhanced Backward Secrecy.
- Utilizes the CTR_DRBG construction specified in NIST SP 800-90A, qualifying it as an NIST-approved DRBG (deterministic random bit generator).

>     - Operates at 256bit security strength.

Reviewer: ```suggestion
    - Operates at 256 bit security strength.
```

(or 256b)

- Support for multiple separate CSRNG instances per IP block.
    - Each instance has its own internal state, control, reseed counters and IO pins.

>     - The number of RNG instances is set via a module parameter

Reviewer: ```suggestion
    - The number of RNG instances is set via a module parameter.
```

>     - The number of RNG instances is set via a module parameter.

Reviewer 2: RNG -> CSRNG (?)

Author: Good point.  Thank you!

> - Software access to a dedicated RNG instance.

Reviewer 2: Maybe also RNG -> CSRNG (here, I'm less sure)

Because RNG is a "loaded term" here I'm trying make sure we are really specific about what we mean.

Author: Will do.

>     - One instance, Instance 0, is always accessible through TLUL registers,

Reviewer: ```suggestion
    - One instance, Instance 0, is always accessible from the bus through device registers,
```

(If you keep the bus name, write it "TL-UL".)

>     - All other instances route to other hardware peripherals (e.g. the key manager, obfuscation engines etc.) and in normal operation these other instances are inaccessible from software.

Reviewer 4: Did we close on CTR_DRBG or is it still open?
Why do we need more than one RNG instance?
Regardless, I assume there is only going to be one AES instance, right?

Author: I believe CTR_DRBG has been the only implementation considered in every conversation I've had.  If there are objections now is probably the right time to voice them though.
We need multiple instances to have protected state for each consumer.  However yes there is only one AES.

Reviewer 2: I would prefer to not hardwire assumptions about there being only 1 AES engine, if at all possible.

As an IP core, I can easily imagine a situation where someone would put want multiple AES cores.  (I realize that's not the first use-case we are targeting, but precluding it would be annoying.)

Reviewer 5: @cdgori as I see the csrng design lay out, the micro-architecture will only use one AES cipher block. Arbiters will be used to select access to the requester(s). Multiple AES block could be used simply by replacing the arbiters with direct connections. This is the trade-off between area, power, latency, etc.

Reviewer 2: Ah, sorry @mwbranstad, I probably have misunderstood the context, I realize now that there are two different AESes we are talking about.

100% agree that there is only the one AES instance in the CSRNG, w/ arbiters as you describe.  I actually can't see needing more than one AES there even in a (speculative) version of OT core with lots of crypto modules, it just seems unlikely to need to generate CSRNG instances with kind of performance.

I thought (wrongly, bias of thinking about the AES client block lately) that this was talking about how many possible AES clients there could be, which was what I was reacting to.

>     - The IP may be configured to support “debug mode” wherein all instances can be accessed by software.

Reviewer 3: should we consider adding an option to make this OTP enabled?

Reviewer 3: oops, it's right under. sorry.

Author: NP!

>       For security reasons this mode may be permanently disabled using fuses.

Author: Sure.

- The IP interfaces with external entropy sources to obtain any required non-determinstic seed material (entropy) and nonces.
    - Requires an external entropy source which is compliant with NIST SP 800-90B, and which also satisfies the requirements for a PTG.2 class physical non-deterministic random number generator as defined in AIS31.
    - Dedicated hardware interface with external entropy source satisfies requirements for get_entropy_input() interface as defined in SP 800-90A.

>     - When needed, utilizes the Block_Cipher_df derivation function (as defined in SP 800-90A) for assembling seed material.

Author: These backticks, are they meant to refer to function calls in our own code base?  These here are abstract calls, which won't correspond to real tasks or functions anywhere (but will find analogs in FSM sequences or interfaces).   I'll make the change, even though I'm a bit confused about the intent in this case.  Could you please explain?

>     - When needed, utilizes the Block_Cipher_df derivation function (as defined in SP 800-90A) for assembling seed material.

Reviewer: Nothing in Markdown has any semantic meaning, it's only styling. Backticks switch to a mono-spaced font, which makes it easier to distinguish plain text from something that is/could be code, a variable, a parameter, etc.

In this case, it's really a matter of preference and fine either way. Use the style you're most comfortable with!

      This allows the use of entropy sources which are not full-entropy (less than one bit of entropy per bit generated).
- Also supports the optional use of personalization strings or other application inputs (e.g. OTP memory values) during instantiation.

> - Assuming a continuously-live entropy source, each instance can also optionally be used as a non-determinstic TRNG (true random number generator--also called a  non-deterministic random bit generator or NRBG in SP 800-90C).

Author: done

    - In this mode, an instance also meets the requirements laid out for a PTG.3 class RNG, the strongest class laid out in AIS31.

>     - Implementation follows the NRBG “Oversampling Construction” approved by SP 800-90C, to meet both CC and FIPS TRNG constructions.

Reviewer 4: Instead of holding multiple instances, would it be possible to reseed after every consumption if the port requires enhanced forward secrecy?

Reviewer 3: i think the reason to have multiple instances is in fact because some might require frequent reseeds and some might not.  Especially for the HW masking consumption case, the thinking is that reseeds can less frequent.

Reviewer 4: 200 kbps entropy can support reseed every few ms. Isn't that frequent enough for most use cases. The only scenario that I see which will require more frequent reseed is DRG.4 support which I am not sure we need and even if so, we can settle for two instances, one with DRG.4 and one without.
Note that per NIST.SP.800-90a,  the maximum number of requests between reseeds (reseed_interval) is 2^48

Author: Hi @zi-v, this seems to be a place where there are room for disagreement between security experts.   If one wants to use this design in the mode you describe here, it can certainly be done by parametrizing the block to only use one instance, and one state vector.   So this proposal supports the use case you describe here.  

By leaving this a parameter we leave our options open.

> - In addition to the approved DRNG node, any instance can also operate in “Fully Deterministic mode”, meaning the seed depends entirely on application inputs or personalization strings.

Reviewer: node -> mode? (Not sure, could be a special term?)

Author: Changed. Thanks for spotting that.

>     - This provides an approved means of key-derivation in FIPS 140-2.

Reviewer 2: If you have it handy, can we be a little more precise with the reference to FIPS 140-2, i.e. what section says this means is approved?

Author: Here it is: 
`This provides an approved means of seed construction in FIPS 140-2 as described in the [FIPS 140-2 Implementation Guidance](https://csrc.nist.gov/csrc/media/projects/cryptographic-module-validation-program/documents/fips140-2/fips1402ig.pdf), section 7.14, resolution point 2(a).`
Note I've generalized this here to seed generation as key generation is not specifically mentioned there, but we do like this mechanism specifically for key generation.

## Description

Though the recommendations in AIS31 are based around broad functional requirements, the recommendations in SP 800-90 are very prescriptive in nature, outlining the exact constructs needed for approval.
Thus the interface and implementation are largely driven by these explicit constructs, particularly the CTR_DBRG construct.

The CSRNG IP consists of four main components:
1. An AES primitive

> 2. The CTR_DRBG state-machine (CTR_DRBG_FSM) which drives the AES primitive, performing the various encryption sequences proscribed in SP 800-90A.

Reviewer 2: proscribed -> prescribed
(proscribe means forbid)

Author: Right. Good catch, thank you.

These include:

    1. **The Derivation Function:**  
       Part of the instantiation and reseed routines, this routine assembles the previous seed material (on reseed only), application inputs, and entropy.
    2. **The Instantiation Routine:**
       Combines application inputs, external entropy and nonce (more entropy) via the derivation function.
    3. **The Reseed Routine:**
       Combines the previous seed material with external entropy to generate a new seed.
    4. **The Generate Routine:**
       Generates up to CSRNG_MAX_GENERATE random bits.
       If called with prediction_resistance_flag, forces a reseed.
     5. **The Update Routine:**

>         Updates the internal state of the DRBG instance after each generate call.

Reviewer 2: "DRBG instance" -> "DRNG instance" (? - I'm not sure - I just noticed it because you use DRNG instance in the next bullet)

Author: Changed.  The fact that CC and NIST use different acronyms is pretty frustrating at times.  Though I've changed this here and  clarified the usage in the first paragraph.

`In this document the terms "DRNG" are "TRNG", are used most generally to refer to deterministic or true random number generation functionalities implemented to this specification.`
`However, the terms "DRBG" or "NRBG" are specifically used when respectively referring to SP 800-90A or SP 800-90C requirements.`
`Meanwhile, when addressing requirements which originate from AIS31 we refer to the specific DRG.3 or PTG.3 classes of RNGs.`

Reviewer 2: Fully agree with your frustration!  That clarification is quite helpful.

Author: Though it has a typo which I just noticed, and will fix.

3. State vectors for each DRNG instance.
4. Interface logic and access control for each instance.

## Note on the term "Entropy"

Every DRNG requires some initial seed material, and the requirements for the generation of that seed material varies greatly between standards, and potentially between CC security targets.
In all standards considered, DRNG's require some "entropy" from an external source to create the initial seed.
However, the rules for obtaining said entropy differ.
Furthermore the required delivery mechanisms differ.
For this reason we must make a clear distinction between "Physical" (or "Live" or "True") entropy and "Factory Entropy".

> This distinction is most important when considering the creation of IP which is both compatible with  IPS 140-3 recommendations, and FIPS 140-2 use cases (which have already been certified for a variety of applications).

Reviewer 2: IPS -> FIPS (unless this is a render issue on my side)

Author: Thanks.

> This distinction is most important when considering the creation of IP which is both compatible with  IPS 140-3 recommendations, and FIPS 140-2 use cases (which have already been certified for a variety of applications).

Reviewer 2: Also, do you have a link to the specific FIPS 140-3 recommendations you are meaning here?

Author: It's just part of SP 800-90, (which are cited in 140-3), but I'll say that explicitly.

- Physical entropy is the only type of "entropy" described in SP 800-90.

> The means of generation is described in SP 800-90B.

Reviewer 2: I'd probably put a ref to the specific 800-90B you are referring to at this spot.

Author: Done

One statistical test requirement is that physical entropy must be unique between reboot cycles, ruling out sources such as one-time programmable (OTP) memories.
In SP 800-90A, the delivery mechanism must come through a dedicated interface and "not be provided by the consuming application".

- Factory entropy is a type of entropy described in the [FIPS 140-2 implementation guidance (IG)](https://csrc.nist.gov/csrc/media/projects/cryptographic-module-validation-program/documents/fips140-2/fips1402ig.pdf).
It can be stored in a persistent memory, programmed at the factory.
In some use cases, the consuming application needs to explicitly load this entropy itself and process it to establish the expected seed.
(However this is permissible in FIPS 140-2.)

This document aims to make the distinction between physical entropy and factory entropy wherever possible.

> However, if used unqualified, the term "entropy" should be understood to refer to physical entropy strings which are obtained in accordance with SP 800-90C.

Reviewer 2: I think this is the spot to link to the specific 800-90C PDF you are referring to.

(As I look at this more, I could be convinced that we could have a final section with external references in this document?  Just to avoid any confusion about which versions we are using, and then the document-internal refs could just point to the references section?  WDYT?)

Author: Made a reference to SP 800-90C Second Draft at the same point my first reference to the term NRBG above.   As far as the links to the references section that sounds nice.  However is there a reference as how to do that in our other Markdown documents?  I'll take a look around.

- Compliant with NIST and BSI recommendations for random number generation.
- In deterministic mode, meets the requirements given in AIS31 for a DRG.3 class deterministic random number generator (DRNG) meaning it provides Forward Secrecy and Enhanced Backward Secrecy.
- Utilizes the CTR_DRBG construction specified in NIST SP 800-90A, qualifying it as a NIST-approved DRBG (deterministic random bit generator).
    - Operates at 256 bit security strength.
- Support for multiple separate CSRNG instances per IP block.
    - Each instance has its own internal state, control, reseed counters and IO pins.
    - The number of CSRNG instances is set via a module parameter.
- Software access to a dedicated CSRNG instance.
    - One instance, Instance 0, is always accessible from the bus through device registers,
    - All other instances route to other hardware peripherals (e.g. the key manager, obfuscation engines etc.) and in normal operation these other instances are inaccessible from software.
    - The IP may be configured to support "debug mode" wherein all instances can be accessed by software.

>       For security reasons this mode may be permanently disabled using one-time programmable (OTP) memory..

Author: Doh!  Thanks for catching that.

- The IP interfaces with external entropy sources to obtain any required non-determinstic seed material (entropy) and nonces.
    - Requires an external entropy source which is compliant with NIST SP 800-90B, and which also satisfies the requirements for a PTG.2 class physical non-deterministic random number generator as defined in AIS31.
    - Dedicated hardware interface with external entropy source satisfies requirements for `get_entropy_input()` interface as defined in SP 800-90A.
    - When needed, utilizes the `Block_Cipher_df` derivation function (as defined in SP 800-90A) for assembling seed material.
      This allows the use of entropy sources which are not full-entropy (less than one bit of entropy per bit generated).

> - Also supports the optional use of personalization strings or other application inputs (e.g. OTP memory values) during instantiation.

Author: Hi @moidx, We _could_, though I hadn't considered it before now. @mwbranstad, This wasn't in our original interface, but would this be hard to add?

> - Also supports the optional use of personalization strings or other application inputs (e.g. OTP memory values) during instantiation.

Reviewer 5: @moidx Yes, all commands support additional data since there is no additional cost to do so.

In all standards considered, DRNG's require some "entropy" from an external source to create the initial seed.
However, the rules for obtaining said entropy differ.
Furthermore the required delivery mechanisms differ.
For this reason we must make a clear distinction between "Physical" (or "Live" or "True") entropy and "Factory Entropy".
This distinction is most important when considering the creation of IP which is both compatible with the both the relatively new SP 800-90 recommendations, as well as the well-established FIPS 140-2 guidelines.

- Physical entropy is the only type of "entropy" described in SP 800-90.
The means of generation is described in [SP 800-90B](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90B.pdf).
One statistical test requirement is that physical entropy must be unique between reboot cycles, ruling out sources such as one-time programmable (OTP) memories.
In SP 800-90A, the delivery mechanism must come through a dedicated interface and "not be provided by the consuming application".

> - Factory entropy is a type of entropy described in the [FIPS 140-2 implementation guidance (IG)](https://csrc.nist.gov/csrc/media/projects/cryptographic-module-validation-program/documents/fips140-2/fips1402ig.pdf) section 7.14, resolution point 2(a).

Reviewer 6: Just to confirm, will software have access to factory_entropy mode in instance 0, Or does this require a second instance?

Author: I may have to defer this to @mwbranstad, but my understanding is that everything in instance 0 is accessible from TL-UL, so you could instantiate that instance in this mode, use it, and then scrap it, and start another instance (in the same instance 0 slot).  However this would be effectively using two instances.  @mwbranstad would it be possible to make multiple instances SW accessible, and then scrap selected ones using OTP?

Reviewer 5: @martin-lueker it turns out that sw interface can start as many instances as are available in the hw. all modes are available to all instances. BTW, due to an interconnect limitation, the sw instance is not 0 now, but the highest number of available APP interfaces. If there are combinations of more than one sw app interface, we would need to discuss this more.

Since these two standards use significantly different terminology, it is recommended that the reader refer to our RNG compliance strategy document for an overview of the various RNG classes and equivalencies between the two standards.
The CSRNG IP supports both of these standards for both deterministic (DRNG) and true random number generation (TRNG).
In NIST terms, it works with the [Entropy Source IP]({{< relref "hw/ip/entropy_src/doc" >}}) to satisfy the requirements as a DRBG (deterministic random-bit-generator) or NRBG (non-deterministic random bit generator).
In AIS31 language, this same implementation can be used to satisfy either the DRG.3 requirements for deterministic generation, or the PTG.3 requirements for cryptographically processed physical generation.

> In this document the terms "DRNG" are "TRNG" are used most generally to refer to deterministic or true random number generation functionalities implemented to this specification.

Author: @imphil: this paragraph is new with this commit.  It would great if you could pound this into shape with your rigorous proofiing skils.
